File: csharp/src/Azpos/AzposCalculator.cs
namespace fgrzl.azpos;

public static class AzposCalculator
{
    private const int MaxLength = 25;

    /// <summary>
    ///     Validates an azpos position string. Throws AzposException if invalid.
    /// </summary>
    public static void Validate(string? pos)
    {
        switch (pos)
        {
            case null:
                throw new AzposException("null pos");
        }
        switch (pos.Length)
        {
            case > MaxLength:
                throw new AzposException("exceeds max length");
        }
        switch (pos)
        {
            case "a":
                throw new AzposException("reserved lower bound");
            case "z":
                throw new AzposException("reserved position");
        }

        foreach (var c in pos)
        {
            switch (c)
            {
                case < 'a' or > 'z':
                {
                    throw c switch
                    {
                        >= 'A' and <= 'Z' => new AzposException("uppercase not allowed"),
                        _ => new AzposException("non-alphabetic character")
                    };
                }
            }
        }
    }

    /// <summary>
    ///     Validates an azpos position string. Returns true if valid, false otherwise.
    /// </summary>
    public static bool IsValid(string? pos)
    {
        try
        {
            Validate(pos);
            return true;
        }
        catch
        {
            return false;
        }
    }

    /// <summary>
    ///     Return lexicographic midpoint between a and b, enforcing a < pos < b
    /// </summary>
    public static string Midpoint(string? a, string? b)
        => MidpointInternal(a, b, true)!;

    private static string? MidpointInternal(string? a, string? b, bool isTopLevel)
    {
        if (a is null || b is null)
            throw new AzposException("null input");

        switch (isTopLevel)
        {
            case true when a == b:
            {
                throw a.Length switch
                {
                    MaxLength => new AzposException("no midpoint"),
                    _ => new AzposException("equal inputs")
                };
            }
        }

        if (Compare(a, b) >= 0)
            throw new AzposException("a must be less than b");

        if (string.IsNullOrEmpty(a))
            return "am";

        if (a.Length < b.Length && a.Length > 0 && b.StartsWith(a, StringComparison.Ordinal))
        {
            var restLen = b.Length - a.Length;
            return restLen > 0 ? $"{a}{new string('a', restLen - 1)}m" : $"{a}m";
        }

        switch (a.Length)
        {
            case 1 when b.Length == 1:
            {
                var diff = b[0] - a[0];
                return diff > 1 ? $"{a}{b}" : diff == 1 ? $"{a}m" : throw new AzposException("no midpoint");
            }
        }

        if (a.Length == b.Length && a.Length > 0)
        {
            var prefix = a[..^1];
            var aLast = a[^1];
            var bLast = b[^1];
            if (a[..^1] == b[..^1] && bLast - aLast == 1)
                return $"{prefix}{a[^1]}m";
        }

        int i = 0, max = Math.Min(a.Length, b.Length);
        while (i < max && a[i] == b[i])
            i++;

        var commonPrefix = a[..i];
        var aSuffix = a[i..];
        var bSuffix = b[i..];

        switch (aSuffix.Length)
        {
            case > 0 when bSuffix.Length > 0:
            {
                var aChar = aSuffix[0];
                var bChar = bSuffix[0];
                var diff = bChar - aChar;

                switch (diff)
                {
                    case > 1:
                    {
                        var midChar = (char)(aChar + diff / 2);
                        return $"{commonPrefix}{midChar}m";
                    }
                    case 1 when aSuffix.Length == 1 && bSuffix.Length == 1:
                        return $"{commonPrefix}{aSuffix[0]}m";
                    case 1 when aSuffix.Length > 1 && bSuffix.Length > 1:
                    {
                        var restA = aSuffix[1..];
                        var restB = bSuffix[1..];
                        if (restA == restB)
                            return $"{commonPrefix}{bSuffix[0]}m";

                        var rec = MidpointInternal(restA, restB, false);
                        return string.IsNullOrEmpty(rec) switch
                        {
                            false => $"{commonPrefix}{aSuffix[0]}{rec}",
                            _ => $"{commonPrefix}{bSuffix[0]}m"
                        };
                    }
                    case 1 when aSuffix.Length != bSuffix.Length:
                        return $"{commonPrefix}{bSuffix[0]}m";
                }

                return isTopLevel switch
                {
                    true => throw new AzposException("no midpoint"),
                    _ => null
                };
            }
        }

        return isTopLevel switch
        {
            true => throw new AzposException("no midpoint"),
            _ => null
        };
    }

    /// <summary>
    ///     Return 3 evenly spaced values between a and c
    /// </summary>
    public static string[] ReBalance(string? a, string? c)
    {
        // For azpos, rebalance3 returns evenly spaced values between a and c, using 'h', 'p', 'x'
        if (a is null || c is null)
            throw new AzposException("null input");
        
        if (a.Length != c.Length)
            throw new AzposException("rebalance3: a and c must be same length");
        
        var baseStr = a;
        return [
            $"{baseStr}h",
            $"{baseStr}p", 
            $"{baseStr}x"
        ];
    }

    /// <summary>
    ///     Return 3 evenly spaced values between a and c (alias for ReBalance)
    /// </summary>
    public static string[] Rebalance3(string? a, string? c) => ReBalance(a, c);

    /// <summary>
    ///     True if midpoint would exceed length or conflict
    /// </summary>
    public static bool NeedsRebalance(string? a, string? b)
    {
        try
        {
            var mid = Midpoint(a, b);
            // Rebalance if midpoint is not valid or if we're getting close to MaxLength
            try
            {
                Validate(mid);
                return mid.Length >= MaxLength - 6;
            }
            catch
            {
                return true;
            }
        }
        catch
        {
            return true;
        }
    }

    /// <summary>
    ///     Lexicographic sort logic (safe in all languages)
    /// </summary>
    public static int Compare(string? a, string? b)
        => string.Compare(a, b, StringComparison.Ordinal);
}

File: java/src/main/java/com/fgrzl/azpos/AzposCalculator.java
package com.fgrzl.azpos;

import java.util.regex.Pattern;

public class AzposCalculator {
    private static final int MAX_LENGTH = 32;
    private static final Pattern VALID_PATTERN = Pattern.compile("^[a-z]*$");
    
    /**
     * Return lexicographic midpoint between a and b, enforcing a < pos < b
     */
    public static String midpoint(String a, String b) throws AzposException {
        return midpointInternal(a, b, true);
    }
    
    private static String midpointInternal(String a, String b, boolean isTopLevel) throws AzposException {
        if (isTopLevel && a.equals(b)) {
            throw new AzposException("equal inputs");
        }
        if (compare(a, b) >= 0) {
            throw new AzposException("a must be less than b");
        }
        
        // If a is empty, always return 'am'
        if (a.isEmpty()) {
            return "am";
        }
        
        // If a is a prefix of b, extend with 'a's and 'm' at the right depth
        if (a.length() < b.length() && b.startsWith(a)) {
            String rest = b.substring(a.length());
            if (rest.length() > 0) {
                return a + "a".repeat(rest.length() - 1) + "m";
            }
        }
        
        // If both are single chars
        if (a.length() == 1 && b.length() == 1) {
            int aChar = a.charAt(0);
            int bChar = b.charAt(0);
            int diff = bChar - aChar;
            if (diff > 1) {
                return a + b;
            }
            if (diff == 1) {
                return a + "m";
            }
        }
        
        // If equal length, all but last char equal, and last char differs by 1
        if (a.length() == b.length() && a.length() > 0) {
            String prefix = a.substring(0, a.length() - 1);
            int aLast = a.charAt(a.length() - 1);
            int bLast = b.charAt(b.length() - 1);
            if (a.substring(0, a.length() - 1).equals(b.substring(0, b.length() - 1)) && bLast - aLast == 1) {
                return prefix + a.charAt(a.length() - 1) + "m";
            }
        }
        
        // Find common prefix
        int i = 0;
        int maxLen = Math.min(a.length(), b.length());
        while (i < maxLen && a.charAt(i) == b.charAt(i)) {
            i++;
        }
        String prefix = a.substring(0, i);
        String aSuffix = a.substring(i);
        String bSuffix = b.substring(i);
        
        // If suffixes are both non-empty and not equal
        if (aSuffix.length() > 0 && bSuffix.length() > 0) {
            int aChar = aSuffix.charAt(0);
            int bChar = bSuffix.charAt(0);
            int diff = bChar - aChar;
            
            // If chars are not adjacent, use midpoint char
            if (diff > 1) {
                char midChar = (char)(aChar + diff / 2);
                return prefix + midChar + "m";
            }
            
            // Characters are adjacent (diff = 1), need to handle carefully
            if (diff == 1) {
                // If both suffixes are length 1, return prefix + aChar + 'm'
                if (aSuffix.length() == 1 && bSuffix.length() == 1) {
                    return prefix + aSuffix.charAt(0) + "m";
                }
                
                // For longer suffixes, prefer the simpler solution when possible
                if (aSuffix.length() > 1 && bSuffix.length() > 1) {
                    String restA = aSuffix.substring(1);
                    String restB = bSuffix.substring(1);
                    
                    // If rest of strings are equal, prefer the shorter solution
                    if (restA.equals(restB)) {
                        return prefix + bSuffix.charAt(0) + "m";
                    }
                    
                    // Try recursion to find midpoint in the rest
                    try {
                        String rec = midpointInternal(restA, restB, false);
                        if (!rec.isEmpty()) {
                            return prefix + aSuffix.charAt(0) + rec;
                        }
                    } catch (AzposException e) {
                        // Recursion failed
                    }
                    
                    // If recursion failed, use bChar + 'm'
                    return prefix + bSuffix.charAt(0) + "m";
                }
                
                // For unequal length suffixes, try to use bChar as base
                if (aSuffix.length() != bSuffix.length()) {
                    return prefix + bSuffix.charAt(0) + "m";
                }
            }
            
            // If we can't find a midpoint
            if (isTopLevel) {
                throw new AzposException("no midpoint");
            }
            return "";
        }
        
        // If we reach here, no valid midpoint can be found
        if (isTopLevel) {
            throw new AzposException("no midpoint");
        } else {
            return "";
        }
    }
    
    /**
     * Ensure position string is lowercase a–z and ≤ max length
     */
    public static boolean validate(String pos) {
        if (pos.length() > MAX_LENGTH) {
            return false;
        }
        
        // Reserved bounds: forbid 'a' and 'z'
        if (pos.equals("a") || pos.equals("z")) {
            return false;
        }
        
        // Check if all characters are lowercase a-z
        return VALID_PATTERN.matcher(pos).matches();
    }
    
    /**
     * True if midpoint would exceed length or conflict
     */
    public static boolean needsRebalance(String a, String b) {
        try {
            String mid = midpoint(a, b);
            // Rebalance if midpoint is not valid or if we're getting close to MAX_LENGTH
            return !validate(mid) || mid.length() >= MAX_LENGTH - 6;
        } catch (AzposException e) {
            return true;
        }
    }
    
    /**
     * Return 3 evenly spaced values between a and c
     */
    public static String[] rebalance3(String a, String c) throws AzposException {
        // For azpos, rebalance3 returns evenly spaced values between a and c, using 'h', 'p', 'x'
        if (a.length() != c.length()) {
            throw new AzposException("rebalance3: a and c must be same length");
        }
        String base = a;
        return new String[]{base + "h", base + "p", base + "x"};
    }
    
    /**
     * Lexicographic sort logic (safe in all languages)
     */
    public static int compare(String a, String b) {
        return a.compareTo(b);
    }
}
